### PR TITLE
MaxPoolWithIndices

### PR DESCRIPTION
**Pull request by Krovatkin**

### CODE REVIEW

File: src/ngraph/runtime/cpu/cpu_emitter.cpp [hunk 1]
```diff
@@ -3006,6 +3046,49 @@ namespace ngraph
                 }
             }
 
+            template <>
+            void CPU_Emitter::EMITTER_DECL(ngraph::op::MaxPoolWithIndicesBackprop)
+            {
+                auto mpb = static_cast<const ngraph::op::MaxPoolWithIndicesBackprop*>(node);
+
+                auto delta_shape = args[1].get_shape();
+                auto out_shape = out[0].get_shape();
```
Comment (jbobba) [inline src/ngraph/runtime/cpu/cpu_emitter.cpp]: unused variables.

File: src/ngraph/runtime/cpu/op/max_pool_with_indices.cpp [hunk 2]
```diff
@@ -0,0 +1,350 @@
...
+    , m_window_movement_strides(window_movement_strides)
+    , m_padding_below(padding_below)
+    , m_padding_above(padding_above)
+{
+    auto& arg_shape = get_input_shape(0);
+
+    //
+    // Make sure arg: NCDi for some Di of rank>0, N != 0, C != 0.
+    //
+    if (arg_shape.size() < 3)
```
Comment (jbobba) [inline src/ngraph/runtime/cpu/op/max_pool_with_indices.cpp]: Can we do something similar to ConvolutionRelu. Create a Maxpool op and that would do all this checking and use that to create a MaxPoolWithIndices.

File: src/ngraph/runtime/cpu/op/max_pool_with_indices.cpp [hunk 3]
```diff
@@ -0,0 +1,350 @@
...
+    //
+    // Construct result shape: NCDo.
+    //
+    Shape result_shape(1 + 1 + spatial_dimension_count);
+    result_shape[0] = batch_size;
+    result_shape[1] = channel_count;
+    copy(output_item_shape.begin(), output_item_shape.end(), result_shape.begin() + 2);
+
+    add_output(get_input_element_type(0), result_shape);
+    add_output(element::i32, result_shape);
```
Comment (jbobba) [inline src/ngraph/runtime/cpu/op/max_pool_with_indices.cpp]: We will need a defintion/comment about the value of indices. Get the definition from mkldnn folks. Depending on that we can also define the element type.

File: src/ngraph/runtime/cpu/op/max_pool_with_indices.hpp [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+        /// 2. <i>(the window movement strides, optional)</i> a vector of positive integers \f$(s_1,\dots,s_n)\f$.
+        ///
+        /// The output has the shape \f$(N,C,d'_1,\dots,d'_n)\f$, where \f$d'_n = \lceil \frac{d_i - w_i + 1}{s_i} \rceil\f$.
+        ///
+        /// Given an input data batch tensor \f$T_\textit{in}\f$, the output tensor is defined by the equation
+        ///
+        /// \f[
+        ///      T_\textit{out}[a,c,i_1,\dots,i_n] = \max_{j_1 = s_1 i_1, \dots, j_n = s_n i_n}^{j_1 = s_1 i_1 + w_1 - 1, \dots, j_n = s_n i_n + w_n - 1} (T_\textit{in}[a,c,j_1,\dots,j_n])
+        /// \f]
+        ///
```
Comment (jbobba) [inline src/ngraph/runtime/cpu/op/max_pool_with_indices.hpp]: Remove and point to documentation in MaxPool op. Dont want to duplicate the comments.

File: src/ngraph/runtime/cpu/pass/cpu_fusion_bprop.cpp [hunk 5]
```diff
@@ -0,0 +1,147 @@
+/*******************************************************************************
```
Comment (jbobba) [inline src/ngraph/runtime/cpu/pass/cpu_fusion_bprop.cpp]: ngraph/runtime/cpu/pass/workspace_insertion.cpp? Probably ask around if folks have better name.

File: src/ngraph/runtime/cpu/pass/cpu_layout.cpp [hunk 6]
```diff
@@ -765,6 +766,79 @@ namespace ngraph
...
+                        node =
+                            insert_input_conversions(external_function, node, prim_input_formats);
+                        set_output_layouts(node, prim_output_formats);
+                    }
+                    else
+                    {
+                        set_default_layouts(external_function, node);
+                    }
+                }
+
```
Comment (jbobba) [inline src/ngraph/runtime/cpu/pass/cpu_layout.cpp]: Reuse code. Look at ConvolutionLayout.

**Comment by Krovatkin** [comment]
> This is also going to need a max_pool_with_indices .rst file. I will be making a PR with max_pool.rst soon.

This is a backend-specific/MKLDNN op. Are we documenting those now?

**Comment by diyessi** [comment]
Then you just need something useful to an IA transformer person in the .hpp file.

**Comment by jbobba** [comment]
Agree with @diyessi  My main point is to not duplicate

File: src/ngraph/runtime/cpu/op/max_pool_with_indices.cpp (hunk 3 above)
Comment (Krovatkin) [reply src/ngraph/runtime/cpu/op/max_pool_with_indices.cpp]: MKLDNN uses either `u8` or `s32` to encode indices , [source](https://github.com/intel/mkl-dnn/blob/2df65197f1c841473144a6fa7e3118628baac773/src/cpu/cpu_pooling_pd.hpp#L35) I'd suggest we could always use `i32` until there's some evidence that we need to optimize for space or to reduce copying.

File: src/ngraph/runtime/cpu/pass/cpu_layout.cpp (hunk 6 above)
Comment (Krovatkin) [reply src/ngraph/runtime/cpu/pass/cpu_layout.cpp]: Will do

File: src/ngraph/runtime/cpu/op/max_pool_with_indices.cpp (hunk 2 above)
Comment (Krovatkin) [reply src/ngraph/runtime/cpu/op/max_pool_with_indices.cpp]: This is actually a **preferred** way as per @aprocter and @diyessi . We kinda cut a few corners for `ConvolutionRelu`. I'd rather change `ConvolutionRelu` to include more checks.

File: src/ngraph/runtime/cpu/cpu_emitter.cpp (hunk 1 above)
Comment (Krovatkin) [reply src/ngraph/runtime/cpu/cpu_emitter.cpp]: Will fix.

File: src/ngraph/runtime/cpu/pass/cpu_fusion_bprop.cpp (hunk 5 above)
Comment (Krovatkin) [reply src/ngraph/runtime/cpu/pass/cpu_fusion_bprop.cpp]: *Loud thudding sounds of Nick's head banging against a keyboard*

File: src/ngraph/runtime/cpu/op/max_pool_with_indices.hpp (hunk 4 above)
Comment (Krovatkin) [reply src/ngraph/runtime/cpu/op/max_pool_with_indices.hpp]: Will do

File: src/ngraph/runtime/cpu/op/max_pool_with_indices.cpp (hunk 3 above)
Comment (jbobba) [reply src/ngraph/runtime/cpu/op/max_pool_with_indices.cpp]: i meant this: https://github.com/intel/mkl-dnn/blob/2df65197f1c841473144a6fa7e3118628baac773/src/cpu/ref_pooling.cpp#L92 
'kh*KW + kw'

TF MaxPoolwithArgMax has a different definition of ArgMax.

**Pull request merged**